src/libtiled/tilelayer.cpp

```diff
@@ -305,28 +316,40 @@ void TileLayer::flipHexagonal(FlipDirection direction)
 
     const char (&flipMask)[16] = (direction == FlipHorizontally ? flipMaskH : flipMaskV);
 
-    for (int y = 0; y < mHeight; ++y) {
-        for (int x = 0; x < mWidth; ++x) {
-            const Cell &source = (direction == FlipHorizontally ? cellAt(mWidth - x - 1, y) : cellAt(x, mHeight - y - 1));
-            Cell &dest = newGrid[x + y * mWidth];
-            dest = source;
-
-            unsigned char mask =
-                    (static_cast<unsigned char>(dest.flippedHorizontally()) << 3) |
-                    (static_cast<unsigned char>(dest.flippedVertically()) << 2) |
-                    (static_cast<unsigned char>(dest.flippedAntiDiagonally()) << 1) |
-                    (static_cast<unsigned char>(dest.rotatedHexagonal120()) << 0);
-
-            mask = flipMask[mask];
-
-            dest.setFlippedHorizontally((mask & 8) != 0);
-            dest.setFlippedVertically((mask & 4) != 0);
-            dest.setFlippedAntiDiagonally((mask & 2) != 0);
-            dest.setRotatedHexagonal120((mask & 1) != 0);
+    QHashIterator<QPoint, Chunk* > it(mChunks);
+    while (it.hasNext()) {
+        it.next();
+        for (int y = 0; y < CHUNK_SIZE; ++y) {
+            for (int x = 0; x < CHUNK_SIZE; ++x) {
+                int _x = it.key().x() * CHUNK_SIZE + x;
+                int _y = it.key().y() * CHUNK_SIZE + y;
+
+                if (!contains(_x, _y))
+                    continue;
+
+                Cell dest(it.value()->cellAt(x, y));
+
+                unsigned char mask =
+                        (static_cast<unsigned char>(dest.flippedHorizontally()) << 3) |
+                        (static_cast<unsigned char>(dest.flippedVertically()) << 2) |
+                        (static_cast<unsigned char>(dest.flippedAntiDiagonally()) << 1) |
+                        (static_cast<unsigned char>(dest.rotatedHexagonal120()) << 0);
+
+                mask = flipMask[mask];
+
+                dest.setFlippedHorizontally((mask & 8) != 0);
+                dest.setFlippedVertically((mask & 4) != 0);
+                dest.setFlippedAntiDiagonally((mask & 2) != 0);
+                dest.setRotatedHexagonal120((mask & 1) != 0);
+
+                (direction == FlipHorizontally) ? newLayer->setCell(mWidth - _x - 1, _y, dest)
+                                                : newLayer->setCell(_x, mHeight - _y - 1, dest);
+            }
         }
     }
 
-    copyGrid(newGrid);
+    mChunks = newLayer->mChunks;
+    delete newLayer;
```

Comment:
I realized this only worked because we were never deleting the chunks. I've fixed that leak by pushing a change where the chunk instances are managed by the QHash.

Btw, instead of manual deletion, please declare the `newLayer` as:
```c++
QScopedPointer<TileLayer> newLayer
```